src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt

```diff
@@ -103,54 +105,117 @@ class BackgroundLinkingIndexingService(
-                                "Unable to un-index linking entity with from batch of linking ids $linkingIds.", ex
-                        )
-                    } finally {
-                        unLock(linkingIds)
-                    }
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun linkingUnIndexing() {
+        if (isLinkingIndexingEnabled()) {
+            executor.submit {
+                try {
+                    logger.info("Starting background linking indexing task.")
+                    val esw = Stopwatch.createStarted()
+                    generateSequence(unIndexCandidates::poll)
+                            .chunked(LINKING_INDEX_SIZE)
+                            .forEach { candidateBatch ->
+                                limiter.acquire()
+                                val (linkingEntityKeyIdsWithLastWrite, linkingIds) = mapCandidates(candidateBatch)
+
+                                try {
+                                    unIndex(linkingEntityKeyIdsWithLastWrite, linkingIds)
+                                } catch (ex: Exception) {
+                                    logger.error("Unable to un-index from batch of linking ids $linkingIds.", ex)
+                                } finally {
+                                    unLock(linkingIds)
+                                    limiter.release()
+                                }
+                            }
+                    logger.info(
+                            "Finished background linking un-indexing task in ${esw.elapsed(TimeUnit.MILLISECONDS)} ms."
+                    )
+                } catch (ex: DistributedObjectDestroyedException) {
+                    logger.error("Linking un-indexing queue destroyed.", ex)
                 }
+            }
+        }
+    }
+
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun updateIndexCandidates() {
+        if (isLinkingIndexingEnabled()) {
+            executor.submit {
+                try {
+                    logger.info("Registering linking ids needing indexing.")
+                    getDirtyLinkingIds().filter {
+                        val expiration = lock(it.second)
+                        if (expiration != null && Instant.now().toEpochMilli() >= expiration) {
+                            logger.info("Refreshing expiration for linking id {}", it.second)
+                            refreshExpiration(it.second)
+                            false
+                        } else expiration == null
+                    }.forEach(indexCandidates::put)
+                } catch (ex: Exception) {
+                    logger.info("Encountered error while updating candidates for linking indexing.", ex)
+                }
+            }
+        }
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun updateUnIndexCandidates() {
+        if (isLinkingIndexingEnabled()) {
```

Comment:
seems like this can also be combined with `updateIndexCandidates` (and same comment about unnesting)